src/proposal.mligo

```diff
@@ -73,6 +79,54 @@ let is_ge_qt(qt_1, qt_2 : quorum_threshold * quorum_threshold): bool =
   let (nat_1, nat_2) = cmp_qt (qt_1, qt_2) in
   nat_1 >= nat_2
 
+[@inline]
+// Multiply two fractions
+let mul_qt(qt_1, qt_2 : quorum_threshold * quorum_threshold): quorum_threshold =
+  { numerator = qt_1.numerator * qt_2.numerator; denominator = qt_1.denominator * qt_2.denominator }
+
+[@inline]
+// Reciprocal of a fractions
+let inv_qt(qt: quorum_threshold): quorum_threshold =
+  { numerator = qt.denominator; denominator = qt.numerator }
+
+[@inline]
+// Adds two fractions
+let add_qt(qt_1, qt_2 : quorum_threshold * quorum_threshold): quorum_threshold =
+  { numerator = (qt_1.numerator * qt_2.denominator) + (qt_2.numerator * qt_1.denominator); denominator = qt_1.denominator * qt_2.denominator }
+
+[@inline]
+// Subtracts two fractions, and throws error if second one is larger.
+let sub_qt(qt_1, qt_2 : quorum_threshold * quorum_threshold): quorum_threshold = let
+  n = match is_nat((qt_1.numerator * qt_2.denominator) - (qt_2.numerator * qt_1.denominator)) with
+    | Some n -> n
+    | None -> ((failwith "UNEXPECTED_NEGATIGE_FRACTION_RESULT") : nat) in
+  { numerator = n; denominator = qt_1.denominator * qt_2.denominator }
+
+let rec hcf_(n1, n2: nat * nat): nat =
+  match (ediv n1 n2 : (nat * nat) option) with
+    | Some n -> if n.1 = 0n then n2 else hcf_(n2, n.1)
+    | None -> ((failwith "DIVISON_BY_ZERO"): nat)
+
+// HCF of two naturals
+let hcf(n1, n2: nat * nat): nat =
+  if (n1 > n2) then hcf_(n1, n2) else hcf_(n2, n1)
+
+[@inline]
+// Normalize a fraction so that the numerator and denominator are smallest
+// possible values.
+let normalize_qt(qt: quorum_threshold): quorum_threshold = let
+  hcf = hcf(qt.numerator, qt.denominator)
+  in {numerator = qt.numerator / hcf; denominator = qt.denominator / hcf }
```

Comment:
Yeah, that may be a good idea too. There's many calculations involving percentages, I'd imagine that often the denominator could be divided by `100` and the nominator kept sufficiently detailed (Ideally however if we can avoid unnecessary multiplications it would be better).

Considering that the denominator needs to always be bigger than the numerator, we can pick one sufficiently large and assume that it's detailed enough to be meaningful.

Reply:
> I'd imagine that often the denominator could be divided by 100 and the nominator kept sufficiently detailed 

I have included a rounding off behavior as you suggest here. Right now it normalizes and rounds off after it. May be only rounding off is enough. What do you think?

Reply:
I think rounding off will be enough, avoiding normalization will likely save a considerable amount of gas.

Reply:
@pasqu4le I have removed the normalization function and only uses a recursive rounding function now. It is recursive because sometimes it appears that one iteration is not enough to bring the pair under the limit. May be it could be made more efficient.

Reply:
@sras I think we should avoid recursive functions and that this can be implemented with a single execution (well enough).

Keeping in mind that the `numerator` is always smaller than the `denominator` we can:
1. calculate by what "factor" we want to reduce the fraction
2. divide both values by that "factor"

I also suggest to increase the max `denominator` to a million (which should still be small enough to not be near to the `nat` upper limit) and that such a number would be precise enough to not need the `numerator` check.

So, my suggestion is to use this:
```
[@inline]
let round(qt: quorum_threshold): quorum_threshold =
  let factor = qt.denominator / 1000000n in
  if factor > 1n
  else { numerator = qt.numerator / factor; denominator = qt.denominator / factor}
  then qt
```

Dealing with `nat` division, in the worst case we can obtain either:
- a `numerator` of one, which is still precise to the 0.0001%
- a `denominator` of 2 millions (or maybe less, not sure how `nat` division rounds) which is still not a problem

What do you think, does that sound correct?